xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +135,743 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+// configuration or not.
+func (b *outlierDetectionBalancer) noopConfig() bool {
+	return b.cfg.SuccessRateEjection == nil && b.cfg.FailurePercentageEjection == nil
+}
+
+func (b *outlierDetectionBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	lbCfg, ok := s.BalancerConfig.(*LBConfig)
+	if !ok {
+		b.logger.Errorf("received config with unexpected type %T: %v", s.BalancerConfig, s.BalancerConfig)
+		return balancer.ErrBadResolverState
+	}
+
+	// Reject whole config if child policy doesn't exist, don't persist it for
+	// later.
+	bb := balancer.Get(lbCfg.ChildPolicy.Name)
+	if bb == nil {
+		return fmt.Errorf("outlier detection: child balancer %q not registered", lbCfg.ChildPolicy.Name)
+	}
+
+	if b.child == nil || b.cfg.ChildPolicy.Name != lbCfg.ChildPolicy.Name {
+		b.childMu.Lock()
+		if b.child != nil {
+			b.child.Close()
+		}
+		b.child = bb.Build(b, b.bOpts)
+		b.childMu.Unlock()
+	}
+
+	b.mu.Lock()
+	b.cfg = lbCfg
+
+	addrs := make(map[string]bool, len(s.ResolverState.Addresses))
+	for _, addr := range s.ResolverState.Addresses {
+		addrs[addr.Addr] = true
+		if _, ok := b.addrs[addr.Addr]; !ok {
+			b.addrs[addr.Addr] = newAddressInfo()
+		}
+	}
+	for addr := range b.addrs {
+		if !addrs[addr] {
+			delete(b.addrs, addr)
+		}
+	}
+
+	var interval time.Duration
+	if b.timerStartTime.IsZero() {
+		b.timerStartTime = time.Now()
+		for _, addrInfo := range b.addrs {
+			addrInfo.callCounter.clear()
+		}
+		interval = b.cfg.Interval
+	} else {
+		interval = b.cfg.Interval - now().Sub(b.timerStartTime)
+		if interval < 0 {
+			interval = 0
+		}
+	}
+
+	if b.intervalTimer != nil {
+		b.intervalTimer.Stop()
+	}
+	if !b.noopConfig() {
+		b.intervalTimer = afterFunc(interval, func() {
+			b.intervalTimerAlgorithm()
+		})
+	} else {
+		// "If a config is provided with both the `success_rate_ejection` and
+		// `failure_percentage_ejection` fields unset, skip starting the timer and
+		// do the following:"
+		// "Unset the timer start timestamp."
+		b.timerStartTime = time.Time{}
+		for _, addrInfo := range b.addrs {
+			// "Uneject all currently ejected addresses."
+			if !addrInfo.latestEjectionTimestamp.IsZero() {
+				b.unejectAddress(addrInfo)
+			}
+			// "Reset each address's ejection time multiplier to 0."
+			addrInfo.ejectionTimeMultiplier = 0
+		}
+	}
+
+	// Inhibit picker updates until run() processes the	lbCfgUpdate. This
+	// makes sure a single picker update gets sent out synchronously as a result
+	// of this UpdateClientConnState call, and this picker update contains the
+	// most updated no-op config bit and most recent state from the child.
+	b.inhibitPickerUpdates = true
+	b.mu.Unlock()
+	b.childMu.Lock()
+	err := b.child.UpdateClientConnState(balancer.ClientConnState{
+		ResolverState:  s.ResolverState,
+		BalancerConfig: b.cfg.ChildPolicy.Config,
+	})
+	b.childMu.Unlock()
+
+	done := make(chan struct{})
+	b.pickerUpdateCh.Put(lbCfgUpdate{
+		lbCfg: lbCfg,
+		done:  done,
+	})
```

Comment:
Why can't we do this inline? Why do we need to go through this channel? If we set `inhibitPickerUpdates` to `true` at the top of this function, any updates from child policies in the meantime would be dropped until we push a picker from here, and reset `inhibitPickerUpdates`.